src/posts/2022-11-27-define-once/define-once.md

```diff
@@ -0,0 +1,355 @@
+cell is run again. Conveniently, the `globals()` dictionary gives us access to
+the global namespace, enabling us to see if a function with a given name already
+exists and hold onto the original definition instead of the new (likely
+equivalent) definition.  The name of a function can be looked up through the
+`__name__` property, so we can define the following decorator[^2] to find the
+existing function and keep it.
+
+[^2]: A decorator is just a higher order function (a function that takes another
+      function as an argument).
+
+```python
+def define_once_first_try(f):
+    if f.__name__ in globals().keys():
+        # Return the original function, ignoring the new
+        # definition of `f`.
+        return globals()[f.__name__]
+    
+    # Otherwise, the function does not already exist and we should
+    # assign this new function to the global namespace.
+    return f 
+```
+
+Now we can decorate our function with `define_once_first_try` once it is
+memoized[^3] to prevent subsequent reruns of a cell from clearing out our cache.
+
+[^3]: The order of decorators actually does not matter; `cache` is smart enough
+      to notice that a cache for the specific function already exists.
+
+```python
+@define_once_first_try
+@cache
+def long_running_notebook_function(*args, **kwargs):
+    # Now our notebook will no longer take forever to run!
+    return result
+```
+
+Now we can go back to merrily coding in our notebook, using "Run All Cells"
+recklessly without worrying about recomputing expensive results.
+
+At some point, we realize our `long_running_notebook_function` has a bug. No
+problem, we can go back and fix the function and go back to the task at hand.
+
+```python
+@define_once_first_try
+@cache
+def long_running_notebook_function(*args, **kwargs):
+    # Now our notebook will no longer take forever to run!
+    
+    # We were missing a constant 5 from our results. Fixed now!
+    return result + 5
+```
+
+We rerun the cell defining `long_running_notebook_function` and all the cells
+that need this function. But we are getting the same resulting buggy results as
+before. Hmm, what is going on?
+
+
+## Since we only define a function once, we can't modify it!
+
+Our strategy for preventing a function from being redefined is a tad too strong:
+it only keeps track of the first definition of the code, meaning we can never
+alter that definition. It would be amazing if we could write perfect code so
+that we would not need to redefine a function, but alas we are human. What the
+`define_once` function should really be doing is detecting _if the function
+meaningfully changes_, and if it does, to accept the new definition
+(invalidating the old cache in the process).
+
+We can get a function's source code using the `inspect.getsource` function from
+the standard library and use the source code to see if the function changed in a
+meaningful way. If the code has not changed, then we should return the original
```

Comment:
Why do you say "in a meaningful way" if you are doing a literal equality comparison between the sources? Is the source compiled in some way at this point such that formatting/commenting/renaming isn't captured?

Reply:
I put "meaningful" in quotes here as a change. My idea was to keep using the term "meaningful change" in the post and then going "wait our definition of meaningful was off, lets try a new one" until I get to the final discussion about function extensionality.